src/nnet3bin/nnet3-train.cc

```diff
@@ -74,10 +79,50 @@ int main(int argc, char *argv[]) {
     {
       NnetTrainer trainer(train_config, &nnet);
 
+      // extract recurrent output names and their offsets from nnet 
```

Comment:
Move all the state-preserving code into a separate function. Ensure that the size of the function is < 40 lines, unless absolutely necessary.